src/irmin/tree.ml

```diff
@@ -622,12 +678,40 @@ module Make (P : Private.S) = struct
             | Some x, Some y -> if equal_node x y then True else False
             | _ -> Maybe)
 
+    let is_updatemap_empty v um =
+      let any_add =
+        StepMap.to_seq um
+        |> seq_exists (function _, Remove -> false | _, Add _ -> true)
+      in
+      if any_add then false
+      else
+        let val_is_empty = P.Node.Val.is_empty v in
+        if val_is_empty then true
+        else
+          let remove_count = StepMap.cardinal um in
+          if (not val_is_empty) && remove_count = 0 then false
+          else (
+            (* This is expensive from this point but there is no alternative *)
+            cnt.node_val_list <- cnt.node_val_list + 1;
+            let entries = P.Node.Val.list v in
+            if List.length entries > remove_count then false
+            else
+              let any_kept =
+                (* Would converting to [Set] and use [subset : t -> t -> bool]
+                   be quicker? *)
```

Comment:
Potentially; this would be faster if the update map is much larger than the node. In practice, I think optimising this is not worthwhile.

The really worthwhile optimisation here would be using a `Val.seq` instead of `Val.list`. https://github.com/mirage/irmin/issues/1257